Validate location by target

kathsherratt opened the pull request:
Validates location with reference to a dataframe of locations by target variable.

This is needed because not all locations have truth data for all target variables, so there are some combinations of target variable/location that are invalid.

This PR:

- Adds code to generate `locations_eu.csv` . 
   - There are currently duplicate `locations_eu.csv` files in both hub and validations repos
   - This needs updating in at least the validations repo: so that locations each have a truth data source listed by target variable, with NA where there is no source of truth data
   - The R code for creating this has been copied across from PR https://github.com/epiforecasts/covid19-forecast-hub-europe/pull/710 and the csv file re-generated
   - This is necessary in this (validations) repo, but not strictly necessary in the main hub repo. So we can continue to merge the [open hub PR](https://github.com/epiforecasts/covid19-forecast-hub-europe/pull/710), which keeps everything in parallel between hub/validations, or close it completely to keep the hub csv simpler (though location csvs would then be out of sync) *.
- Edits `project_variables.py` so that locations are read in as full dataframe (and location column extracted wherever needed)
- Modifies step 1 of the row-by-row validation in `covid19.py`
   - Both target variable and location detected for each row
   - Location dataframe filtered to non-NA value for that location/target variable combination
   - Location in the forecast row must be in the filtered locations dataframe, otherwise error
   - Error message includes the problematic target variable as well as location.

I've tested all locally, and set up our temporary-testing PR 536 with new targets (code to create csv [here](https://gist.github.com/kathsherratt/277bf15204f3ff11eab0c0b21ff27b4f)), so we can re-trigger the validations there when this is in main.

 [* We will certainly need to edit this system of having duplicate locations files, which will be easy to do when we switch to using R for all the validation]

kathsherratt committed:
add the locations template using R

kathsherratt committed:
Update locations_eu.csv

with relevant source by location for each target

kathsherratt committed:
validate each location with reference to target

sbfnk commented:
Super! Just one question:

> Adds code to generate `locations_eu.csv` .

Would it be easier to read this directly from the other repo? It seems possible to pass a url to `read_csv` https://stackoverflow.com/questions/32400867/pandas-read-csv-from-url

kathsherratt committed:
read locations direct from main hub csv

using url of main hub: data-locations/locations_eu.csv

kathsherratt committed:
use project locations (from url) in test_formatting.py

kathsherratt committed:
remove hardcoded locations

not needed as now downloaded from main hub repo in project_variables.py and imported from there elsewhere

kathsherratt commented:
Yep, much more sensible! Tested locally again just to double check and all works.

kathsherratt commented:
Merging in order to test before opening up submissions to `inc hosp` targets.
